[CR]Make monsters and NPCs regen outside bubble

Coolthulhu opened the pull request:
`monster` and `npc` classes get following new members:
- `last_updated`
- `on_unload` - currently only resets the above. Later on it may cache some extra info, for example NPC could cache that there's a nearby bed
- `on_load` - updates the critter according to difference between `last_updated` and `calendar::turn`

Currently only makes monsters regenerate when loaded. This regeneration first checks quick regeneration (shoggoth), then revival, finally critter's material (flesh and veggy heals). The best of those is selected.
Regeneration works at full speed offscreen, zeds regen ~1 hp per hour, fleshy and veggy stuff ~1 hp per 4 hours.

[CR] because healing rates for non-regenerating stuff may need adjustments and because zombies may also want to heal lost speed.

[WiP] because I'd like to add sleeping NPC regen (including tiredness removal) here too.

DanmakuDan commented:
Just wondering, would self-revived zombies heal up to full?

BevapDin commented:
Looks good.

It could be used to fast-forward some of the AI effects of NPCs, those that control when you asked them for an item or similar.

It appears `unload_npcs(); load_npcs();` is quite often used in that combination, maybe put it into anadditional function `reload_npcs` or so, not sure if it's worth it.

Coolthulhu commented:
> would self-revived zombies heal up to full

Not exactly full. Full hp yes, but not full speed.
This speed part is one of the reasons why it's [CR] - I wasn't sure if it should regen or not and how fast.
I'd probably put it at some lower rate than hp, to make it so that zombies first "scar over" and only then properly regenerate.

> It could be used to fast-forward some of the AI effects of NPCs, those that control when you asked them for an item or similar.

Yeah, timing out effects will probably go here.
For monsters, the best way would probably be to do multiple passes, each ending with a damaging effect timing out.
Say, a zombie has: "onfire": 3, "bleed": 10 and 10 days of updating. When updating, it would first update 3 turns, during which it would take damage from fire and bleed, then 7 turns for bleed, then (if still alive) it would get 10 days (without 10 turns) of regen.

For NPCs it would be a bit more complex, as the NPC would consume bandages instead of dying, but other than that it should be doable in a similar way.

Rivet-the-Zombie commented:
Very nice!

Coolthulhu commented:
This grew much bigger than I expected, because getting NPCs to regenerate properly required a giant refactor of a bunch of functions.
It will be a good thing once I get it to work, because now functions are grouped much better. Fatigue regeneration near tiring (before, some of it was near pain, other near sleep), sleep hp regeneration in own function, made `get_effect` return a reference instead of copy.

Did some minor rebalancing, changed the way some values are quantized (statistically the effect should be similar) making it clearer how does "bio_recycler" compare against LIGHTEATER (spoiler: 4 times less food required with bio), changed broken part mending to tick every 30 minutes (rather than 100 minutes) and rebalanced mend chance to account for it.

Overall the balance shouldn't be affected much.
I'd appreciate if someone could point out what I copy+pasted badly, where I inverted logic etc. The PR is big and I probably got some things wrong.

kevingranade commented:
I feel like this particular feature is a huge pain to test from within the game, and will be very easy to write some unit tests (integration tests probably, but whatever) for.
The tests would just look like "spawn a monster/NPC, fiddle with their stats, unload the map they're on, advance the calendar, load the map they're on, examine their stats".

kevingranade commented on src/game.cpp:
```diff
@@ -763,9 +764,31 @@ void game::load_npcs()
                 temp->my_fac->known_by_u = true;
-            active_npc.push_back(temp);
+            active_npc.push_back( temp );
+            just_added.push_back( active_npc[active_npc.size() - 1] );
```
You can replace ```active_npc[active_npc.size() - 1]``` with ```active_npcs.back()```, but why not just use temp?

kevingranade commented on src/monster.cpp:
```diff
@@ -2053,3 +2053,34 @@ void monster::hear_sound( const tripoint &source, const int vol, const int dist
+{
+    // Possible TODO: Integrate monster upgrade
+    const int dt = calendar::turn - last_updated;
+    last_updated = calendar::turn;
```
I think you did the right thing here, but I just wanted to bring it up.
The only time dt will be less than 0 is if we debugged the calendar value (pretty much don't care about things being consistent there) or if the monster is loaded in a new game.  When interacting between saves at different times, we pretty much have two options, either monsters will advance too rapidly, or they won't advance at all sometimes.  This does the first, if you move ```last_updated = calendar::turn;``` under the if conditional, we'd get the second behavior instead.
I think what you've done is the right thing, because I expect the dominant use case to be "play one game for a while, die, start new game" rather than "toggle back and forth between two games".
Might want a comment here outlining that we're making that tradeoff.

kevingranade commented on src/monster.cpp:
```diff
@@ -2053,3 +2053,34 @@ void monster::hear_sound( const tripoint &source, const int vol, const int dist
+    const int heal_amount = divide_roll_remainder( regen * dt, 1.0 );
+    const int healed = heal( heal_amount );
+    add_msg( m_debug, "on_load() by %s, %d turns, healed %d", name().c_str(), dt, healed );
+}
```
nit, git likes having newlines at ends of files.

kevingranade commented on src/player.cpp:
```diff
@@ -5467,169 +5501,243 @@ void player::update_needs()
+    const bool has_recycler = has_bionic("bio_recycler");
+    const bool asleep = in_sleep_state();
+    const bool hibernating = asleep && is_hibernating();
+    float hunger_rate = 1.0f;
```
Not a must-have, particularly for this PR, but it'd be nice to extract these guys to player::hunger_rate(), plater::fatigue_rate(), etc.

kevingranade commented:
Whew, what a monster.
I gave it an once-over and nothing jumps out at me as problematic.  Considering its size, I'd prefer to get it landed without any more features if possible, or perhaps extract some of the non-functional API changes to a seperate PR we can land, just so this guy gets smaller.
The regeneration rates you've chosen look good to me.  I can see regenerating speed, either that or use HP as a proxy, if HP hits max, just reset the monster instance and reset their stats.
Fatigue removal for NPCs doesn't sound too complicated, but if it turns out to be, please do punt to a new PR instead of struggling on with this one.

Coolthulhu commented:
Thanks for the review.

Added a test to compare retroactive vs. turn-by-turn. Currently looks like retroactive isn't perfect, but I'd say it's good enough of an approximation for NPC-only stuff.

Hacked around retroactively updated sleeping NPCs. This is a minor hack, which very weakly affects the player: a sleeping player who gets the "you feel well rested" will wake up a tiny bit sooner after that than expected (after 2 minutes from "well rested" rather than 3-5 minutes).

Activating debug mode before loading critters will say how long were they "away" and in case of monsters how much did they regen.

NPC retroactive update is capped at 2 days and processed in 30 minute intervals, then 5 minute ones (with the last 30 minute one being divided into 6*5 minutes), the single turns.

Effect decay (for stuff like asking NPCs for equipment and for sleep) postponed because too huge PR.

kevingranade commented on tests/npc_test.cpp:
```diff
@@ -0,0 +1,79 @@
+    CHECK( who.fatigue >= -25 );
+}
+
+TEST_CASE("NPC on_load correctly approximates hunger/thirst/fatigue")
```
The main thing these test case names are used for is running just this test, like ```tests/cata_test "NPC on_load correctly approximates hunger/thirst/fatigue"```, so it's nicer to have them be slightly shorter and cli-friendly, like "npc-needs-update-on-load".
Seperate issue, if the test doesn't pass, you can make it not run by default by passing a second argument to TEST_CASE() which is a string starting with a period, including just a period.

kevingranade commented on src/player.h:
```diff
@@ -198,14 +198,25 @@ class player : public Character, public JsonSerializer, public JsonDeserializer
-        /** Increases hunger, thirst, fatigue, stimms wearing off, dying from hunger and dying from overdose */
-        void update_needs();
-        /** Handles passive regeneration of pain and maybe hp, except sleep regeneration.
+        /** Updates all "biology" as if time between `from` and `to` passed, or by one turn */
```
I'm gathering from the structure here that update_body() *must* be called every turn for the player, and void update_body( int from, int to ) is used to process a span of time having passed for a NPC, that needs to be documented here.  If I'm wrong, it's even important to document how it works here :)

kevingranade commented on tests/npc_test.cpp:
```diff
@@ -0,0 +1,79 @@
+        on_load_test( test_npc, 0, MINUTES(5 * five_min_ticks) );
+
+        const int margin = 10;
+        CHECK( test_npc.get_hunger() <= five_min_ticks + margin );
```
Hunger/thirst rate is cut in half while sleeping.

kevingranade replied:
Verified locally, if the predicted hunger/thirst is "number of 5-minute time periods" / 2, it falls within the margin.
I suspect the remaining discrepancies are similar.

Coolthulhu commented:
Yeah, it was sleep - good catch.

That sleep breaks one of the tests though - sleep fatigue rate depends on sleep time, meaning testing fatigue regeneration won't work until retroactive effect processing does.

I managed to get the rest of the tests to work, though. Added a test that simulates 6 turns iterating one-by-one and using `on_load` and it passes too.

kevingranade commented:
Woo, victory of unit tests!
